# Introduction

  ATPS is a computer program to automate the positive selection pipeline which fetches sequences of species to scan for positively selected genes based on sequence comparison and codon-based method. The program mainly deals with different software which operates together by python coding. It handles the input and output files in each step and saves the results in directory for each gene.
  we have developed ATPS to detects positively selected genes (PSGs), allows analysis of specific evolutionary branches and also it print out the Bayes Empirical Bayes (BEB) analysis which figure out the positive amino acid sites. ATPS allow user to fetch data from NCBI or input your data using arguments.  

![GitHub all releases](https://img.shields.io/github/downloads/APS-P/APSP/total?color=%2300ff00&label=Downloads&logo=GitHub&logoColor=white&style=plastic) / ![GitHub Pipenv locked Python version](https://img.shields.io/github/pipenv/locked/python-version/APS-P/APSP?label=Python&logo=github&style=plastic)

# Wiki (How to use the program)

A detailed documentation including: How to use the tool, its arguments, and a detailed description of what's included can be found at the [GitHub Wiki](https://github.com/APS-P/APSP/wiki). In short, it contains a detailed guide on how to use ATPS

# Dependencies

   # softwares 
   
1- MUSCLE

2- MAFFT

3- clustalo

4- Gblocks

8- jmodeltest

9- phyml

8- phylofit

9- codeml

   # python packages 
   
1- biopython

2- pandas

3- glob

4- matplotlib 

5- os 

6- bokeh

7- mne

8- numpy

9- gc 

10- tkinter 


# Technical support

ATPS is actively supported by our team. Please use the [Issues section](https://github.com/APS-P/APSP/issues) if you happen to find any errors.

## Citation
Will be added after publication and official release.
